Packs/CommonScripts/Scripts/FormatURL/FormatURL.py

```diff
@@ -1,291 +1,549 @@
+        self.base = 0
+        self.output = ''
+
+        self.inside_brackets = False
+        self.port = False
+        self.query = False
+        self.fragment = False
+        self.done = False
+
+        self.sub_delims = ("!", "$", "&", "'", "(", ")", "*", "+", ",", ";", "=")
+        self.brackets = ("\"", "'", "[", "]", "{", "}", "(", ")", ",")
+        self.url_code_points = ("\u0021", "\u0024", "\u0026", "\u0027", "\u0028", "\u0029", "\u002A", "\u002B",
+                                "\u002C", "\u002D", "\u002E", "\u002F", "\u003A", "\u003B", "\u003D", "\u003F",
+                                "\u0040", "\u005F", "\u007E")
+
+        if self.original_url:
+            self.remove_leading_chars()
+
+        else:
+            raise URLError("Empty string given")
+
+        if "//" in self.modified_url[:8]:
+            """
+            The URL seems to have a scheme indicated by presence of "//"
+            """
+            self.scheme_check()
+
         try:
-            # IPv6 should have square brackets around them, no need to remove
-            ipaddress.IPv6Address(re.findall(r"\[(.*?)]", url)[0])
-            return True
+            first_slash = self.modified_url[self.base:].index("/")
 
-        except (ipaddress.AddressValueError, IndexError):
-            return False
+        except ValueError:
+            first_slash = -1
+
+        try:
+            if "@" in self.modified_url[:first_slash]:
+                """
+                Checks if url has '@' sign in its authority part
+                """
+                self.user_info_check()
+
+        except ValueError:
+            # No '@' in url at all
+            pass
+
+        self.host_check()
+
+        if not self.done and self.port:
+            self.port_check()
+
+        if not self.done:
+            self.path_check()
+
+        if not self.done and self.query:
+            self.query_check()
+
+        if not self.done and self.fragment:
+            self.fragment_check()
+
+    def __str__(self):
+        return f"{urllib.parse.unquote(self.output)}"
+
+    def __repr__(self):
+        return f"{urllib.parse.unquote(self.output)}"
+
+    def scheme_check(self):
+        """
+        Parses and validates the scheme part of the URL, accepts ascii and "+", "-", "." according to standard.
+        """
+
+        index = self.base
+        scheme = ''
+
+        while self.modified_url[index].isascii() or self.modified_url[index] in ("+", "-", "."):
+            """
+            Scheme State checker
+            """
+
+            char = self.modified_url[index]
+            if char in self.sub_delims:
+                raise URLError(f"Invalid character {char} at position {index}")
+
+            elif char == ":":
+
+                self.output += char
+                index += 1
+
+                if self.modified_url[index:index + 2] != "//":
+                    "If URL has ascii chars and ':' with no '//' it is invalid"
```

Comment:
Is this supposed to be a comment?
```suggestion
                    # If URL has ascii chars and ':' with no '//' it is invalid
```

Reply:
yep :)